bench_test.go

```
@@ -0,0 +1,179 @@
+		require.NoError(b, err)
+		require.Equal(b, http.StatusAccepted, resp.Result().StatusCode, string(respBody))
+		measurements = append(measurements, time.Since(start))
+	}
+
+	if print {
+		printBenchmark("PUT benchmark", measurements)
+	}
+	return mhs
+}
+
+func getMultihashes(b *testing.B, mhs []multihash.Multihash, server *http.ServeMux, print bool) {
+	measurements := make([]time.Duration, 0, len(mhs))
+	for _, mh := range mhs {
+		start := time.Now()
+		resp := httptest.NewRecorder()
+		req := httptest.NewRequest(http.MethodGet, "/multihash/"+mh.B58String(), nil)
+		server.ServeHTTP(resp, req)
+
+		respBody, err := io.ReadAll(resp.Body)
+		require.NoError(b, err)
+		require.Equal(b, http.StatusOK, resp.Result().StatusCode, string(respBody))
+		measurements = append(measurements, time.Since(start))
+	}
+	if print {
+		printBenchmark("GET benchmark", measurements)
+	}
+}
+
+func printBenchmark(name string, m []time.Duration) {
+	var total int64
+	var max int64
+
+	for _, d := range m {
+		total += int64(d)
+		if int64(d) > max {
+			max = int64(d)
+		}
+	}
+
+	mean := int64(total) / int64(len(m))
+	var deviation int64
+
+	for _, d := range m {
+		deviation += int64(math.Pow(float64(int64(d)-mean), 2))
+	}
+	deviation = int64(math.Sqrt(float64(deviation / int64(len(m)))))
+
+	log.Print(name)
+	log.Printf("\tmean: %v", time.Duration(mean))
+	log.Printf("\tstdev: %v", time.Duration(deviation))
+	log.Printf("\tmax: %v", time.Duration(max))
+}
+
+func createServer(b *testing.B, dir string) (*dhstore.PebbleDHStore, *metrics.Metrics, *http.ServeMux) {
+	store, err := dhstore.NewPebbleDHStore(dir, nil)
+	require.NoError(b, err)
+
+	m, err := metrics.New("0.0.0.0:40081")
+	require.NoError(b, err)
+
+	return store, m, dhstore.NewHttpServeMux(store, m)
+
+}
+
+func randomMultihashes(b *testing.B, n int) []multihash.Multihash {
+	mhs := make([]multihash.Multihash, n)
+
+	bytes := make([]byte, 100)
+	for i := 0; i < n; i++ {
+		crand.Read(bytes)
```

* Use deterministic seeded RNG, and take it as a parameter for reproducible benchmarks.
* error is not handled on this line.

---

I changed it to crypto rand because using the math/rand failed the static check. Let me revert it back.

---

The main thing is to use a seeded RNG for reproducibility. 

The `staticcheck` error you saw I suspect is because of the recent deprecation of default RNG offered by `rand` package, (which is also non-deterministic).

---

Should I use the one from `math/rand`?

---

Something like this: `rng := rand.New(rand.NewSource(<seed-goes-here>))`, and make sure:
*  the `rng` is instantiated once and passed around to whatever function that needs it
* is accessed sequentially.